### ISSUE TITLE
Language guessing: Allow paste contents to override `files.defaultLanguage`?

### ISSUE THREAD

**Issue by JacksonKearl** [microsoft/vscode#148093]
Speaking with @sbatten it was griped that the "paste contents into an untitled editor to set the language to the paste source" flow doesn't work when you have `files.defaultLanguage` configured. This is intentional AFAIK but maybe worth revisiting? Relevant code:

https://github.com/microsoft/vscode/blob/33c614d11a561bfac4061cdb5f24dc4215c62e9e/src/vs/workbench/browser/parts/editor/textResourceEditor.ts#L165-L220

cc @TylerLeonhardt @bpasero

**Comment by sbatten** [comment]
For my reasoning, I have had the default language setting to markdown forever as I use Ctrl+N to create random notes often. However, that doesn't mean I always want to have a markdown file. It seems the setting is given too much precedence even though its a very non-contextual setting. If I paste JSON, and I haven't typed anything else in this editor, in my experience, its never been right to keep the markdown.

**Comment by TylerLeonhardt** [comment]
I think @Tyriar initially convinced me to honor `files.defaultLanguage` in language detection with the reasoning that someone who sets `files.defaultLanguage` knows enough about VS Code to figure out how to set their language to something else when they need it, while the real customer of the feature was new users who didn't know how to do that.

**Comment by bpasero** [comment]
I also notice how when I copy text from e.g. a TypeScript file and paste that into a new untitled file, we do not set the language based on the source, but simply use `files.defaultLanguage`. I would argue in that case we should ignore the setting and use the source language because we know the language in that case.

**Comment by sbatten** [comment]
+1 to @bpasero's comments.

As for what @TylerLeonhardt said, just because I know how to change the language doesn't mean there aren't cases where its annoying that I have to. Currently, I feel punished for using the default language setting, because everywhere we make language detection nicer, I'm left out because I take notes in markdown and wanted that to be the default language.

**Comment by bpasero** [comment]
I think the problem is somewhere here:

https://github.com/microsoft/vscode/blob/21718208c101d77d5d07382f954e227c94203760/src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts#L193-L202

This method in the end will set `_hasLanguageSetExplicitly`:

https://github.com/microsoft/vscode/blob/c8745c3a9cddb9b22ce16524236c08583f1bf1b6/src/vs/workbench/common/editor/textEditorModel.ts#L84

And based on that flag, we skip auto detection.

I guess we have to decide what condition we want for `_hasLanguageSetExplicitly` to be `true` or `false`:
* explicit language change by the user from status bar: probably yes
* `files.defaultLanguage`: probably not
* other ways in which via API a untitled file is opened with explicit language: need to decide case by case

We probably would have to transport this information down to the untitled model to set the flag accordingly.

**Comment by TylerLeonhardt** [comment]
> As for what @TylerLeonhardt said, just because I know how to change the language doesn't mean there aren't cases where its annoying that I have to. Currently, I feel punished for using the default language setting, because everywhere we make language detection nicer, I'm left out because I take notes in markdown and wanted that to be the default language.

This would have to be configurable. I had this the default behavior and got a handful of folks complaining saying "vscode changed the language on me when I didn't want it to"

**Comment by sbatten** [comment]
Configurable is fine, but I'm curious if those cases truly match mine or they are being lumped together because we weren't fine-grained about our approach. I am not arguing that anytime we auto-detect the language we should change it on the user.

@JacksonKearl's hint bulb is a great improvement, but pasting text we know the language for seems a much stronger signal than just language detection

**Comment by TylerLeonhardt** [comment]
Is there a way to differentiate between pasted code and non-pasted code? Could we run the "to be pasted" code through language detection and if it feels strongly enough, we set the language in addition to pasting the content?

**Comment by sbatten** [comment]
in the base case of this issue, we know exactly the language because it comes from an existing file. We don't rely on the model in this case.

**Comment by bpasero** [comment]
Yeah we do, the language detection [here](https://github.com/microsoft/vscode/blob/33c614d11a561bfac4061cdb5f24dc4215c62e9e/src/vs/workbench/browser/parts/editor/textResourceEditor.ts#L166-L168) is specifically running from a paste operation.

**Comment by Tyriar** [comment]
👍 to configurable, I paste far more into untitled markdown files than anything else and sometimes I start with pasting code before wrapping it in backticks. Since I'm comfortable with the change language keybinding I prefer to do that in the off chance I need it. If we do make it configurable I'd suggest just making a separate setting and not have the language detection code consider the default language setting at all.
